c++/src/kj/compat/http.h

```diff
@@ -718,16 +718,18 @@ class HttpClient {
   // stack-allocated).
 
   struct ConnectResponse {
-    uint statusCode;
-    // Any 2xx response indicates that the CONNECT request was successful without regard to the
-    // standard semantics of the specifix 2xx code.
-    kj::StringPtr statusText;
-    const HttpHeaders* headers;
-    kj::OneOf<kj::Own<kj::AsyncInputStream>, kj::Own<kj::AsyncIoStream>> connectionOrBody;
-    // If the CONNECT request is successful, the connectionOrBody will be a kj::AsyncIoStream
-    // through which both directions of the tunnel are accessible. If the CONNECT request errors,
-    // then the connectionOrBody will be a kj::AsyncInputStream used to provide the payload
-    // (if any) of the error.
+    struct Status {
+      uint statusCode;
+      kj::StringPtr statusText;
+      const HttpHeaders* headers;
+      kj::Maybe<kj::Own<kj::AsyncInputStream>> errorBody;
+    };
+
+    kj::Promise<Status> status;
+    // If the CONNECT request is successful, the status will indicate a 2xx status code
+    // and the errorBody will be nullptr. If the CONNECT request fails, then the errorBody
+    // will be a kj::AsyncInputStream used to provide the payload (if any) of the error.
+    kj::Own<kj::AsyncIoStream> connection;
   };
 
   virtual kj::Promise<ConnectResponse> connect(kj::StringPtr host, const HttpHeaders& headers);
```

Comment:
Shouldn't this return `ConnectResponse` now, without the outer Promise?

(Maybe we rename it to `ConnectRequest` since it's very similar to the `Request` type that `request()` returns?)

Reply:
It certainly could. The only reason it's returning a promise is the `httpOutput.flush()` but we can switch it around to return a `PromisedAsyncIoStream` that waits on that

Reply:
This ends up being a much larger change than what I wanted to make here. If we want to change this signature, my preference is to do that as a separate PR.

Reply:
How so? It looks like the biggest thing is you need the write direction on `AsyncIoStreamWithDeferredInput` to be blocked by its own guard promise (which waits for `httpOutput.flush()`), but that's not that complicated?

I'd rather group all the breaking API changes together here...

Reply:
It's not a matter of complexity as much as size of the change. Changing the signature touches a lot more code than is strictly necessary to enable the pipelining. But ok. I'll have to work on it next week tho.

Reply:
This change already changes the signature, though? Wouldn't it touch the same code that was already touched?